### PR TITLE
Wlr 0.17

### PR DESCRIPTION
**Pull request by jwijenbergh**
This patch adds wlroots 0.17 support

### CODE REVIEW

**Comment by elParaguayo** [comment]
@tych0 @ramnes Tagging you both here to say @jwijenbergh is hoping to have this ready to merge by the weekend.

Arch are looking to package `pywlroots 0.17` which was just released but are happy to hold off for a bit while we don't support it. So, I'd suggest we do next release once this PR is merged.

**Comment by tych0** [comment]
Sure, I will have some time to do a release this weekend if we land it by then.

**Comment by jwijenbergh** [comment]
This pretty much finished but the CI is segfaulting, probably similar to the issue we had (and never resolved) with pywlroots > 0.16.4

**Comment by tych0** [comment]
If you think it's the same failure as before and you're otherwise happy with the branch, I'd say go ahead and merge it and let's do a release.

**Comment by jwijenbergh** [comment]
> If you think it's the same failure as before and you're otherwise happy with the branch, I'd say go ahead and merge it and let's do a release.

It seems like it's working now after bumping `pywayland`, let me wait for full CI

**Comment by jwijenbergh** [comment]
Should be ready now, bit difficult to test everything but I have been daily driving this for a while without any real issues. There are some small TODOs for upcoming releases though, will make issues for those

**Comment by jwijenbergh** [comment]
release workflow is failing:

image

I don't understand

**Comment by jwijenbergh** [comment]
caused by:

```
Traceback (most recent call last):
  File "/__w/qtile/qtile/builder.py", line 50, in build_wheel
    import wlroots  # noqa: F401
    ^^^^^^^^^^^^^^
  File "/tmp/build-env-uwvjz0ui/lib/python3.12/site-packages/wlroots/__init__.py", line 8, in <module>
    from ._ffi import ffi, lib
ImportError: while loading wlroots._ffi: failed to import ffi, lib from xkbcommon._ffi

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/python/cp312-cp312/lib/python3.12/site-packages/pyproject_hooks/_in_process/_in_process.py", line 373, in <module>
    main()
  File "/opt/python/cp312-cp312/lib/python3.12/site-packages/pyproject_hooks/_in_process/_in_process.py", line 357, in main
    json_out["return_val"] = hook(**hook_input["kwargs"])
                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/python/cp312-cp312/lib/python3.12/site-packages/pyproject_hooks/_in_process/_in_process.py", line 271, in build_wheel
    return _build_backend().build_wheel(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/__w/qtile/qtile/builder.py", line 52, in build_wheel
    import wlroots
  File "/tmp/build-env-uwvjz0ui/lib/python3.12/site-packages/wlroots/__init__.py", line 8, in <module>
    from ._ffi import ffi, lib
ImportError: while loading wlroots._ffi: failed to import ffi, lib from xkbcommon._ffi```

**Comment by tych0** [comment]
This is the thing where you need to run ffibuild, right? But also, I have not tested the release scripts either. I guess this is from the test pypi upload job?

**Comment by jwijenbergh** [comment]
> This is the thing where you need to run ffibuild, right? But also, I have not tested the release scripts either. I guess this is from the test pypi upload job?

It's this step https://github.com/jwijenbergh/qtile/blob/141da0adcab6dfe6328e908f18198fbcb4b57ad9/.github/workflows/release.yml#L202, haven't changed anything there

**Comment by tych0** [comment]
That almost looks to me like the ffi build for xkbcommon isn't getting run as part of the wheel build... I have never really understood how all that fits together, and a brief look didn't reveal anything. I'll take a look later.

How are you actually running this? it looks like on a successful run, it will really publish the stuff to pypi given that it's not the test job that's failing, but the real one.

**Comment by elParaguayo** [comment]
The issue seems to be the 1.0 release of xkbcommon. I changed the release workflow to this and it worked:
```
python -m pip install cffi build auditwheel cairocffi xcffib pywayland setuptools-scm setuptools wheel
python -m pip install xkbcommon==0.8
python -m pip install --no-binary :all: pywlroots
python -m build -v --config-setting backend=wayland --no-isolation --wheel .
```
I think the `--no-binary` is not necessary.

**Comment by jwijenbergh** [comment]
> That almost looks to me like the ffi build for xkbcommon isn't getting run as part of the wheel build... I have never really understood how all that fits together, and a brief look didn't reveal anything. I'll take a look later.
> 
> How are you actually running this? it looks like on a successful run, it will really publish the stuff to pypi given that it's not the test job that's failing, but the real one.

I am running this by creating a release in my fork. So I expect the upload to fail but not the build

**Comment by jwijenbergh** [comment]
Ok. The problem is that the manylinux do not have a recent enough xkbcommon: https://almalinux.pkgs.org/8/almalinux-powertools-x86_64/libxkbcommon-x11-devel-0.9.1-1.el8.x86_64.rpm.html (needs 1.0)

**Comment by jwijenbergh** [comment]
I disabled pypy builds completely. Now all good

image

File: .github/workflows/ci.yml
Comment (elParaguayo) [inline .github/workflows/ci.yml:20]: Are we all agreed that we're now dropping pypy 3.10 (as opposed to just allowing it to fail)?

No objections from me!

**Review by elParaguayo — APPROVED** [review]
Great work in getting this done!

File: .github/workflows/ci.yml
Comment (jwijenbergh) [reply .github/workflows/ci.yml:20]: My thought process was that for releases it would be nice that CI doesn't fail. After the release we can re-enable it

Comment (tych0) [reply .github/workflows/ci.yml:20]: Fine with me, though can you rebase and drop the continue-on-error I just added as well?

Comment (tych0) [reply .github/workflows/ci.yml:20]: If we do get some movement on that issue, we can always drop the note from the CHANGELOG and re-add pypy to CI.

Comment (tych0) [reply .github/workflows/ci.yml:20]: Oh, derp. You had it!

**Comment by elParaguayo** [comment]
OK - this breaks a fair bit of stuff on qtile-extras. I'll need to work through changes to get it working!

**Comment by tych0** [comment]
> Edit: nowhere near as bad as I feared. All sorted!

heh, did qtile migrate help you out? :)

**Pull request merged**
